File: DSA-C++/2-maths/8-frequency-of-elements.cpp
#include <bits/stdc++.h>

using namespace std;

// brute force - O(n²), O(n)
vector <pair <int,int>> freq_of_elements_1(vector <int> nums) {
    vector <pair <int,int>> ans;
    int n = nums.size();
    vector <bool> visited(n, false);

    for (int i=0; i<n; ++i) {
        int cnt = 1;

        if (visited[i]) {
            continue;
        }

        for (int j=i+1; j<n; ++j) {
            if (nums[i] == nums[j]) {
                visited[j] = true;
                cnt++;
            }
        }

        ans.push_back({nums[i], cnt});
    }

    return ans;
}

// hashing - O(n), O(n)
unordered_map <int,int> freq_of_elements_2(vector <int> nums) {
    unordered_map <int,int> umap;

    for (int x:nums) {
        umap[x]++;
    }

    return umap;
}

int main() {
    vector <int> nums = {10,5,10,15,10,5};

    vector <int> ans = freq_of_elements_1(nums);
    
    // vector <int> ans = freq_of_elements_2(nums);

    for (auto [k,v]:ans) {
        cout << k << "->" << v << endl;
    }

    cout << endl;
    return 0;
}

File: DSA-Python/3-maths/8-frequency-of-elements.py
# Given an array, print the number of occurences of each element in array



# brute force - O(n²), O(n)
def frequency_of_elements_1(nums: list) -> None:
    n = len(nums)
    visited = [False] * n
    
    for i in range(n):
        cnt = 0
            
        if visited[i] == True:
            continue
        
        for j in range(i, n):
            
            if nums[i] == nums[j]:
                visited[j] = True
                cnt += 1
        
        print(nums[i], '=>', cnt)
        
    print()
    
    
# hashing - O(n), O(n)
def frequency_of_elements_2(nums: list) -> None:
    umap = {}
    
    for x in nums:
        umap[x] = umap.get(x, 0) + 1
        
    for (k, v) in umap.items():
        print(k, '=>', v)
        
    print()
    
    
if __name__ == "__main__": 
    n = int(input())
    
    for i in range(n):
        nums = list(map(int, input().split()))
        
        # frequency_of_elements_1(nums)
        
        frequency_of_elements_2(nums)
        
        
'''
3
1 2 3 4

3 3 3 3 -1 -1 2
'''